### DISCUSSION TITLE
Can the maximum current slider be changed?

### DISCUSSION THREAD

**Discussion by lawrencemaroun** [lbbrhzn/ocpp#391] [Q&A] [answered]
The maximum current slider only goes up to 32a. I have a 40a charger and I can't set it to 40a.

**Comment by dreimer1986** [comment, marked as answer]
Maybe it would be an idea to add a max current value to the initial configuration.
I have a 16A charger for example and setting more causes a SuspendedEV state.
This would allow you to select 40A as maximum and me to select 16A.
As this is a number entity and the max is just set to 32 by default, it can be set to something else, too.
At least I was able to hackfix it with the following entry in my customize.yaml:

```
number.evbox_maximum_current:
  max: 16
```

Optional as not needed, but nice for showing what you are setting up in the UI:
```
  unit_of_measurement: A
```

![image](https://user-images.githubusercontent.com/29841368/153885465-af2ac90c-c211-4b0a-aec3-7225b70dd2e2.png)

So if you do the same with 40 as max value and the correct entity name, it might help you until there is another fix.

Sadly someone over at Home Assitant had the questionable idea to remove the customize panel in 12.2021 because it was not perfect...
It would have made things easier for the non hardcore freaks dreaming in yaml by now ^^

EDIT:

Likely in your case there might be a limit in api.py, too.

```
    async def set_charge_rate(self, limit_amps: int = 32, limit_watts: int = 22000):
```
@ line 527 looks like it could be a problem for your case. Maybe change that to 40 is working, maybe not...

And while you are at it you of course could go into numbers.py @ line 34...

```
NUMBERS: Final = [
    OcppNumberDescription(
        key="maximum_current",
        name="Maximum_Current",
        icon=ICON,
        initial_value=32,
        min_value=0,
        max_value=32,
        step=1,
    ),
]
```
... and could change the max value here, too.

**Reply by drc38** [reply]
@dreimer1986 are you going to work up a PR now you've got one under your belt?

**Reply by lawrencemaroun** [reply]
@dreimer1986 Thank you! I edited numbers.py and that worked. But when I update to a new version of the OCPP integration, won't my numbers.py file get overwritten and I will need to do it again?

**Reply by dreimer1986** [reply]
> @dreimer1986 are you going to work up a PR now you've got one under your belt?

Not really. I hope I did not step on someone's toe now... If I did, I did not mean to and send out a sincere apology. I try to understand the code, but this is a few levels above my skill level. I have no clue if my assumption works at all and can't test it, too. (You aren't even able to get a 40 A station over here) And what I posted is not more than a hack. I am by far not fluent enough in Python to create a real solution where you can set it on first time configure. In my opinion that would be a solution and not my stupid limit increase. Especially not for others who have a even more current limited charging solution. 

My problem is that I start typing and then think about it later and then add more and more and even more. (That's why all my stuff is edited half a dozen times.) And in this case I suddently was checking the source instead of just telling about customize.yaml I was sort of sure that my suggestion would cause nothing useful in the case of a limit increase instead of my decrease and then we had a quite long result close to a untested hackfix.

@lawrencemaroun exactly that will happen if you want to stay up to date and I strongy suggest to do that. It's not more than a short lived experiment, unless you wanna change the code again and again. The real solution has to come from the code here and not from a hack in a discussion.  ^^ If it works though, I suggest you open this as a (feature request) issue. There is a button for that on the right.

Btw, personal question. What charger has 40 A?? This is crazy. Our local power supplier would chop off.... things off my body if I would try to use such a charger.

**Reply by lawrencemaroun** [reply]
@dreimer1986 what country are you in? I am in Canada. I've heard of chargers with as much as 80a, but to have one of those your home needs to have 320a service. Most older homes here only have 100a service (like mine did) and I had to upgrade to 200a (which is the standard now) before I could get a 40a charger. We also may not be using the same voltages and phase count as you. Level 2 chargers here are all single phase 240v, so 40a amounts to 9.6kw. My current car only has a 6.6kw (28a) onboard charger so the 32a limit is not an immediate concern for me. My next car will be able to use the full 40a though so I want to get this fixed by then.

I'll go ahead and submit the feature request. I've been looking through the code and will be attempting to contribute new features in the future. It's a bit hard though because there is not much documentation. You seem to know the code pretty well, have you just been analyzing it?

**Reply by lbbrhzn** [reply]
Solved in v0.4.21. @lawrencemaroun, can you please check if it works up to 200a?

**Reply by lawrencemaroun** [reply]
@lbbrhzn Thank you so much! It looks like it works, and I can only test up to 40a.

**Reply by lbbrhzn** [reply]
ok, thanks for checking!

**Reply by dreimer1986** [reply]
@lawrencemaroun Germany. We use three phases with 230V here. (Some use one only) Normally we have 16A on three phases which is about 11kW. Some have 22kW, too aka 32A on all three phases.

@lbbrhzn thanks for the real fix. I can confirm that reducing is working. ^^

**Reply by lawrencemaroun** [reply]
@dreimer1986 wow, I did not realize how much more power you can carry with three phases. Even considering that there is a neutral conductor, 4 conductors carrying 16a each seems intuitively like it would carry the same amount of power as two conductors carrying 32a. But 230v at 32a is only 7.3kw and yet you can carry 11kw with the same amount of copper if you use three phase vs single phase.

**Reply by drc38** [reply]
@dreimer1986 the comment was one of encouragement to develop a fix in the code base ;) This is my first project coding python and HA integrations so it can't be too difficult/complex... @lawrencemaroun if you've got any questions on the code just ask. If you haven't already downloaded a copy the best place to start understanding the integration is the official ocpp and HA developer documentation.

**Reply by dreimer1986** [reply]
@drc38 OK, then I understood it correctly before I started questioning if I interpreted it right :P I will not vanish, that for sure. Maybe I can send in some more things as soon as I understand more of the code and find something I would like to have or see.
